sys/dev/drm/core/drm_ioctl.c

```diff
@@ -0,0 +1,1008 @@
+
+/**
+ * drm_ioctl - ioctl callback implementation for DRM drivers
+ * @filp: file this ioctl is called on
+ * @cmd: ioctl cmd number
+ * @arg: user argument
+ *
+ * Looks up the ioctl function in the DRM core and the driver dispatch table,
+ * stored in &drm_driver.ioctls. It checks for necessary permission by calling
+ * drm_ioctl_permit(), and dispatches to the respective function.
+ *
+ * Returns:
+ * Zero on success, negative error code on failure.
+ */
+long drm_ioctl(struct file *filp,
+	      unsigned int cmd, uintcap_t arg)
+{
+	struct drm_file *file_priv = filp->private_data;
+	struct drm_device *dev;
+	const struct drm_ioctl_desc *ioctl = NULL;
+	drm_ioctl_t *func;
+	unsigned int nr = DRM_IOCTL_NR(cmd);
+	int retcode = -EINVAL;
+	char stack_kdata[128] __aligned(sizeof(void *__capability));
+	char *kdata = NULL;
+	unsigned int in_size, out_size, drv_size, ksize;
+	bool is_driver_ioctl;
+
+	dev = file_priv->minor->dev;
+
+	if (drm_dev_is_unplugged(dev))
+		return -ENODEV;
+
+	is_driver_ioctl = nr >= DRM_COMMAND_BASE && nr < DRM_COMMAND_END;
+
+	if (is_driver_ioctl) {
+		/* driver ioctl */
+		unsigned int index = nr - DRM_COMMAND_BASE;
+
+		if (index >= dev->driver->num_ioctls)
+			goto err_i1;
+		index = array_index_nospec(index, dev->driver->num_ioctls);
+		ioctl = &dev->driver->ioctls[index];
+	} else {
+		/* core ioctl */
+		if (nr >= DRM_CORE_IOCTL_COUNT)
+			goto err_i1;
+		nr = array_index_nospec(nr, DRM_CORE_IOCTL_COUNT);
+		ioctl = &drm_ioctls[nr];
+	}
+
+	drv_size = _IOC_SIZE(ioctl->cmd);
+	out_size = in_size = _IOC_SIZE(cmd);
+	if ((cmd & ioctl->cmd & IOC_IN) == 0)
+		in_size = 0;
+	if ((cmd & ioctl->cmd & IOC_OUT) == 0)
+		out_size = 0;
+	ksize = max(max(in_size, out_size), drv_size);
+
+#ifdef __linux__
+	DRM_DEBUG("pid=%d, dev=0x%lx, auth=%d, %s\n",
+		  task_pid_nr(current),
+		  (long)old_encode_dev(file_priv->minor->kdev->devt),
+		  file_priv->authenticated, ioctl->name);
+#else
+	DRM_DEBUG("%s: pid=%d, auth=%d, %s\n", __func__, task_pid_nr(current),
+	        file_priv->authenticated, ioctl->name);
+#endif
+
+	/* Do not trust userspace, use our own definition */
+	func = ioctl->func;
+
+	if (unlikely(!func)) {
+		DRM_DEBUG("no function\n");
+		retcode = -EINVAL;
+		goto err_i1;
+	}
+
+	if (ksize <= sizeof(stack_kdata)) {
+		kdata = stack_kdata;
+	} else {
+		kdata = kmalloc(ksize, GFP_KERNEL);
+		if (!kdata) {
+			retcode = -ENOMEM;
+			goto err_i1;
+		}
+	}
+
+// Some weird stuff is happening in copy_from_user
+#ifdef __linux__
+	if (copy_from_user(kdata, (void __user *)arg, in_size) != 0) {
+		retcode = -EFAULT;
+		goto err_i1;
+	}
+#else
+	memcpy_c(PTR2CAP(kdata), (void * __capability)arg, in_size);
+#endif
+
+	if (ksize > in_size)
+		memset(kdata + in_size, 0, ksize - in_size);
+
+	retcode = drm_ioctl_kernel(filp, func, kdata, ioctl->flags);
+
+#ifdef __linux__
+	if (copy_to_user((void __user *)arg, kdata, out_size) != 0)
+		retcode = -EFAULT;
+#else
+	memcpy_c((void * __capability)arg, PTR2CAP(kdata), out_size);
```

Comment:
drm_fstub_ioctl is what calls this, taking a `void *` and casting to a `uintcap_t`. For purecap kernels this works, but for hybrid kernels `arg` will be a null-derived capability so the memcpy_c will fault. Either drm_fstub_ioctl needs to do PTR2CAP before the cast or drm_ioctl needs to take a uintptr_t instead (relying on it always being a kernel pointer, and not like in the general linuxkpi equivalent where it can be a user pointer). In isolation the latter is better, as if something is always a kernel pointer then making it clearly that avoids confusion and CHERI specialness, but there's an argument to be made for making it look more like the full-blown linuxkpi version where it can be a user pointer.